components/metris/internal/gardener/handlers.go

```diff
@@ -0,0 +1,125 @@
+	fselector := fields.SelectorFromSet(fields.Set{fieldSecretBindingName: newSecret.Name}).String()
+
+	shootlist, err := c.gclientset.CoreV1beta1().Shoots(newSecret.Namespace).List(metav1.ListOptions{FieldSelector: fselector})
+	if err != nil || len(shootlist.Items) == 0 {
+		c.logger.Debugf("could not find shoots using secret '%s': skipping.", newSecret.Name)
+		return
+	}
+
+	for i := range shootlist.Items {
+		shoot := shootlist.Items[i]
+		c.logger.Debugf("secret '%s' has been modified, updating provider client config for shoot '%s'.", newSecret.Name, shoot.Name)
+		c.shootAddHandlerFunc(&shoot)
+	}
+}
+
+func (c *Controller) shootDeleteHandlerFunc(obj interface{}) {
+	var (
+		shoot *gcorev1beta1.Shoot
+		ok    bool
+	)
+
+	if shoot, ok = obj.(*gcorev1beta1.Shoot); !ok {
+		// try to recover deleted obj
+		delobj, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			c.logger.Error("error decoding object, invalid type")
+			return
+		}
+
+		shoot, ok = delobj.Obj.(*gcorev1beta1.Shoot)
+		if !ok {
+			c.logger.Error("error decoding deleted object, invalid type")
+			return
+		}
+	}
+
+	shootname := fmt.Sprintf("%s--%s", shoot.Namespace, shoot.Name)
+
+	c.logger.Warnf("shoot '%s' is being remove from the provider accounts", shootname)
+
+	c.accountsChan <- &Account{Name: shootname}
+}
+
+func (c *Controller) shootAddHandlerFunc(obj interface{}) {
+	if shoot, ok := obj.(*gcorev1beta1.Shoot); ok {
+		pname := shoot.Spec.CloudProfileName
+		if pname == "az" {
+			pname = "azure"
+		}
+
+		if c.providertype == pname {
+			accountid := shoot.GetLabels()[labelAccountID]
+
+			subaccountid, ok := shoot.GetLabels()[labelSubAccountID]
+			if !ok || subaccountid == "" {
+				// if subaccount label is empty, try the annotation ?
+				subaccountid, ok = shoot.GetAnnotations()[annotationSubAccountID]
+				if !ok {
+					c.logger.Warnf("could not find label '%s' in Shoot '%s', skipping", labelSubAccountID, shoot.Name)
+					return
+				}
+			}
+
+			// secret, err := c.secretInformer.Lister().Secrets(shoot.Namespace).Get(shoot.Spec.SecretBindingName)
+			secret, err := c.kclientset.CoreV1().Secrets(shoot.Namespace).Get(shoot.Spec.SecretBindingName, metav1.GetOptions{})
+			if err != nil {
+				c.logger.Errorf("error getting shoot secret: %s", err)
+				return
+			}
+
+			tenantName, ok := secret.GetLabels()[labelTenantName]
+			if !ok {
+				c.logger.Warnf("could not find label '%s' in secret '%s', skipping", labelTenantName, secret.Name)
+			}
+
+			shootname := fmt.Sprintf("%s--%s", shoot.Namespace, shoot.Name)
+
+			c.logger.With("account", subaccountid).Debug("sending account to provider")
+			c.accountsChan <- &Account{
+				Name:           shootname,
+				ProviderType:   c.providertype,
+				AccountID:      accountid,
+				SubAccountID:   subaccountid,
+				TechnicalID:    shoot.Status.TechnicalID,
+				TenantName:     tenantName,
+				CredentialName: secret.Name,
+				CredentialData: secret.Data,
+			}
+		}
+	} else {
```

Comment:
Just a suggestion. I wonder if it would not be cleaner to write it like this:
```go
shoot, ok := obj.(*gcorev1beta1.Shoot)
if !ok {
    c.logger.Errorf("error getting shoot object from cache")
    return
}
// the rest of the logic...
```
But I guess it's up to you.